File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& nums, int days) {
        int low = *max_element(nums.begin(), nums.end());
        int high = accumulate(nums.begin(), nums.end(), 0);
        int n = nums.size();
        
        while(low <= high) {
            int mid = (low + (high - low) / 2);
            int load = 0, daysNeeded = 1;
            for(int i = 0; i < n; i++) {
                if(load + nums[i] > mid) {
                    daysNeeded++;
                    load = 0;
                } 
                load += nums[i];
            }
            if(daysNeeded <= days) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(nums, days) {
    let n = nums.length;
    let low = Math.max(...nums);
    let high = 0;

    for(var i = 0; i < n; i++) {
        high += nums[i];
    }
    
    while(low <= high) {
        let daysNeeded = 1, load = 0
        let mid = Math.floor((low + high) / 2);
    
        for(let j = 0; j < n; j++) {
            if(load + nums[j] > mid) {
                daysNeeded++;
                load = 0
            }
            load = load + nums[j];
        }
    
        if(daysNeeded > days) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    
    return low;
};